File: android/src/main/java/com/webicon/favicon/FavIconParser.kt
package com.webicon.favicon
import org.jsoup.Jsoup
import java.net.URL

class FavIconParser {


    class FavIconParserImpl {
        fun favIcons(htmlData: ByteArray, baseUrl: URL): List<Favicon> {
            val html = htmlData.toString(Charsets.UTF_8).ifEmpty {
                htmlData.toString(Charsets.ISO_8859_1)
            }

            val document = Jsoup.parse(html, baseUrl.toString())
            val links = document.head().select("link[rel~=icon]") ?: return emptyList()

            return links.mapNotNull { element ->
                val rel = element.attr("rel")
                val href = element.attr("href")
                val sizes = element.attr("sizes")

                if (href.isBlank()) return@mapNotNull null

                val type =
                    if (href.contains(".ico")) Favicon.FavIconType.FAV_ICO else Favicon.FavIconType.from(
                        rel
                    )
                val size = parseSize(sizes)
                val iconUrl = URL(baseUrl, href).toString()

                Favicon(iconUrl, size, type)
            }
        }

        private fun parseSize(sizes: String?): Pair<Int, Int> {
            val parts = sizes?.split("x") ?: return Pair(0, 0)
            return if (parts.size >= 2) {
                Pair(parts[0].toIntOrNull() ?: 0, parts[1].toIntOrNull() ?: 0)
            } else Pair(0, 0)
        }
    }
}

File: ios/FavIconParser.swift
//
//  FavIconParser.swift
//  web
//
//  Created by Sergei Golishnikov on 01/06/2025.
//

import Foundation
import Fuzi

final class FavIconParserImpl {
  func favIcons(from htmlData: Data, relativeTo baseUrl: URL) -> [Favicon] {
    do {
      guard let html = String(data: htmlData, encoding: .utf8)
              ?? String(data: htmlData, encoding: .windowsCP1251),
            !html.isEmpty else {
        return []
      }
      
      let document = try HTMLDocument(string: html, encoding: .utf8)
      return document.head?.children(tag: "link")
        .filter { $0.attr("rel")?.lowercased().contains("icon") == true }
        .compactMap { [weak self] element -> Favicon? in
          return self?.favIcon(from: element, relativeTo: baseUrl)
        } ?? []
    } catch {
      return []
    }
  }
  
  // MARK: - Private
  
  private func favIcon(from element: XMLElement, relativeTo baseUrl: URL) -> Favicon? {
    let relAttribute = element.attr("rel")!
    let sizeAttribute = element.attr("sizes")
    guard var href = element.attr("href") else { return nil }
    let type: Favicon.FavIconType
    if href.contains(".ico") == true {
      type = .favIco
    } else {
      type = Favicon.FavIconType(value: relAttribute)
    }
    let sizeComponents = sizeAttribute?.components(separatedBy: "x")
    var size = CGSize.zero
    if let components = sizeComponents,
       components.count >= 2,
       let width = Int(components[0]),
       let height = Int(components[1]) {
      size = CGSize(width: width, height: height)
    }
    
    if href.starts(with: "//") { href = "https:\(href)"}
    guard let url = URL(string: href, relativeTo: baseUrl) else {
      return nil
    }
    
    return Favicon(url: url, size: size, type: type)
  }
}
